### ISSUE TITLE
RayCaster restructuring

### ISSUE THREAD

**Issue by timoropinski** [inviwo/inviwo#220]
Hi all,

I would like to get feedback on how to restructure our raycasters. I suggest to have three Raycaster processors for now:

1) SimpleVolumeRaycaster:
Properties: As SimpleRaycaster, but with basic Shading functionality

2) VolumeRaycaster (or shall it be called AdvancedVolumeRaycaster?)
Properties: As VolumeRaycaster, but with following changes:
- Rendering: Remove this Property and set to DVR+ISO by default
- Classification: Remove "VoxelValue" entry (or when is this used?)
- Compositing: Remove one of the two "First hit (point)" options
- Gradient: Rename to "Gradient Computation"
- Adjust camera on data change: move out of camera property and rename to "Center view on volume" (analog to MeshRenderer)
- Toggle Shading: Remove (or when is this used?)
- Position Indicator: Rename to "Integrated Geometry" (Rename "Plane 1" .. "Plane 3" to "Geometry 1" to "Geometry 3"); Set Normals by default to be orthogonal; Remove cross option (or does this work?)

3) MultiChannelVolumeRaycaster: as VolumeRaycaster but with 4 Channels

I realize that this requires some network version changes, but I believe it is worth it, as these processors are very central to Inviwo. Do you have any comments or suggestions?

**Comment by martinfalk** [comment]
My two cents:
1) extending `SimpleVolumeRaycaster` with shading functionality will cause quite some changes making it no longer as trivial. And thereby counteracting its purpose as "entry-level" processor.
2) Explanation for `Volume Raycaster` properties
- the `Rendering` property serves the purpose of quickly switching between `DVR`, `Iso`, and `DVR + Iso` modes without having to modify the TF. Otherwise it will for example not be possible to easily hide all isosurfaces
- the classification modes `none`, `tf`, and `voxel` have quite distinctive meanings. `none` passes through the currently selected channel, i.e. results in `vec4(sampledVoxel[channel])` while `tf` will apply the TF to `sampledVoxel[channel]`. In contrast, `voxel` mode will output the sampled voxel with all channels.
See https://github.com/inviwo/inviwo/blob/b5b2602d17b6c0ab7cb516e3acb940624d7ba427/modules/opengl/shader/shaderutils.cpp#L378-L398

- Compositing: the isosurface could be removed
- toggle shading allows to enable/disable shading via a keyboard shortcut. I am using that one quite often. In particular in cases where illumination/shading distorts the shown colors.

**Comment by petersteneteg** [comment]
What about the lightningraycaster?

**Comment by petersteneteg** [comment]
Should also consider the multipel lightsource support that we have in a PR herr...

**Comment by r-englund** [comment]
In research-dev there is a processor that can set the camera to face any side of the bounding box of a volume and make sure the whole bounding box is visible https://github.com/inviwo/research-dev/blob/master/modules/experimental/processors/cameraautoadjuster.cpp#L113-L151

**Comment by petersteneteg** [comment]
this should also be addressed https://github.com/inviwo/inviwo/issues/118
and #58 and #53

**Comment by petersteneteg** [comment]
The position indicator renaming make no sense. Why would you call it "geometry 1" that says nothing about what it is?

**Comment by petersteneteg** [comment]
Question? will not the volume raycaster be pure subset of the multichannelcaycaster? why do we need both?

**Comment by timoropinski** [comment]
@Martin:
1. I agree, but since this will be the processor in most (simple) test workspaces, it should show proper lighting. We could go with light source position, and set all other values to default?!

2. Properties
- Switching between showing and not showing isosurfaces should IMHO happen in the TF property and not on the global property scale. People will thus easier see the connection.
- Classification mode: the I suggest renaming. voxel is also a 'none' mode, but with 4 channels instead of one. So should be 'none (selected channel)' and 'none (all channels)' or similar.
- I would still argue for removing the toggle of the shading, as I believe that there are subjective preferences of what should be toggled, and thus the selection becomes difficult.

@Peter:
- Yes, we should also have an additional illumination raycaster.
- And I would also agree on merging multichannelraycaster and raycaster (it is just three additional TFs, which could only be exposed when selecting all channels).
- Position indicator renaming makes sense, as it can be a plane or a cross at the moment. The properties for each indicator are now always called plane. Also when a cross is selected.

@Rickard: I agree, we should reuse the auto-adjuster functionality (also in the GeoemtryRenderer).

**Comment by ResearchDaniel** [comment]
> SimpleVolumeRaycaster:
> Properties: As SimpleRaycaster, but with basic Shading functionality

I have not strong opinion on this, unless you need it for educational purposes I would leave it as simple as possible, i.e. no shading.

> VolumeRaycaster (or shall it be called AdvancedVolumeRaycaster?)

I think that it can still be called VolumeRaycaster since it should be the go-to choice.

> Rendering: Remove this Property and set to DVR+ISO by default

Agree 

> Classification: Remove "VoxelValue" entry (or when is this used?)

Agree with your previous comment about none (all channels)

> Compositing: Remove one of the two "First hit (point)" options

Agree, they seem to be duplicates? 

> Gradient: Rename to "Gradient Computation"

No objection

> Adjust camera on data change: move out of camera property and rename to "Center view on volume" (analog to MeshRenderer)

Agree

> Toggle Shading: Remove (or when is this used?)

Don't have a strong opinion. Never used it 

> Position Indicator: Rename to "Integrated Geometry" (Rename "Plane 1" .. "Plane 3" to "Geometry 1" to "Geometry 3"); Set Normals by default to be orthogonal; Remove cross option (or does this work?)

I think that Position Indicator is a bit more descriptive than integrated geometry.
Agree on that they should be orthogonal by default. The names could be according to axis (XY, XZ and so on)? 
Cross option did not work for me either

> MultiChannelVolumeRaycaster: as VolumeRaycaster but with 4 Channels

I guess we would like to separate just because multichannel data should have one TF per channel?
I think that we can separate it unless we add/remove TFs depending on the number of channels in the input data. Adding/removing can be a bit dangerous though since it might lead to unwanted removal of a TF.

**Comment by petersteneteg** [comment]
@ResearchDaniel We only need to hide the unused TFs not remove them...

**Comment by petersteneteg** [comment]
@timoropinski about the use if the SimpleRaycaster, do you witth "test workspacces" mean our example workspaces? I think we should use our default VolumeRaycaster in the example workspaces. The SimpleRaycaster is only used as a coding example in our example module.

**Comment by timoropinski** [comment]
Yes, you are right, I mean the example workspaces. Anyway, there I think it would be better, to use the SimpleVolumeRaycaster (at least) in the more simple workspaces. The VolumeRaycaster has too many properties, which impact the learning curve for new users.

**Comment by ResearchDaniel** [comment]
If we could add comments in the workspace it would be possible to point things out. Say for example "SimpleRaycaster is for educational purposes. Use VolumeRaycaster in practice". 

I guess we could even call it EducationalRaycaster...

**Comment by martinfalk** [comment]
@ResearchDaniel see suggestion here https://github.com/inviwo/inviwo-dev/issues/1442
